### ISSUE TITLE
Stream.flatMap has an overly-restrictive type signature

### ISSUE THREAD

**Issue by ceedubs** [unisonweb/base#101]
This applies to both `Stream.flatMap` and `Stream.flatMap!`, so I'll use `Stream.flatMap!` as the example.

The current implementation is:

```
Stream.flatMap! : (a ->{e, Stream b} r) -> '{e, Stream a} r ->{e, Stream b} r
Stream.flatMap! f s = pipe! s '(abilities.forever '(f Ask.ask))
```

However, the `r` that's returned by `f` is totally ignored. Without changing the implementation it could be a generic `r2` or `()` to make it a bit more explicit that its result is inconsequential.

Imagine for example that you want to implement `filter!` in terms of `flatMap!`:

```
up.Stream.filter!' : (a ->{e} Boolean) -> '{e, Stream a} r -> {e, Stream a} r
up.Stream.filter!' pred = Stream.flatMap! (a -> if (pred a) then Stream.emit a else ())
```

Right now this will give the following error:

```
  I found a value of type:  r
  where I expected to find:  Unit
```

However, if we were to change the `flatMap!` signature to the following (without changing the implementation), then it would work just fine.

```
Stream.flatMap!' : (a ->{e, Stream b} r2) -> '{e, Stream a} r -> {e, Stream b} r
Stream.flatMap!' f s = pipe! s '(abilities.forever '(f Ask.ask))
```

**Comment by ceedubs** [comment]
@anovstrup weighed in and suggested using `r2` instead of `()` so the caller doesn't have to throw in an `ignore` if their function doesn't return `()`. Sounds fine to me.

**Comment by ceedubs** [comment]
Resolved by #103
